### DISCUSSION TITLE
How do I prevent content items to be saved?

### DISCUSSION THREAD

**Discussion by intellimedhu** [OrchardCMS/OrchardCore#11402] [Q&A] [answered]
I have a module where I create new versions from content items when users save them.

In another module, there is a content part handler watching the `Publishing` event. My goal is to prevent publishing in some cases, so I tried to throw an exception, but it is swallowed [[see here]](https://github.com/OrchardCMS/OrchardCore/blob/247457e63b2531e3288f017ee95f182a1231bd5b/src/OrchardCore/OrchardCore.Abstractions/Modules/Extensions/InvokeExtensions.cs#L274)

Another way: I set `Context.Cancel`, but it seems more dangerous, because both versions will be unpublished. No exception, no futher information.

Should I write a custom "handler" (avoid deriving from ContentPartHanlder) to solve this issue?

**Comment by hishamco** [comment]
You can use `PublishContentContext.Cancel` with a certain condition to allow only one or some users to publish the content

**Comment by intellimedhu** [comment]
@hishamco 
quoting myself: "I set `Context.Cancel`, but it seems more dangerous, because both versions will be unpublished. No exception, no futher information"
where `Context` is `PublishContentContext`

So I don't want the new version, but the previous and it must be Latest and Published.

(sorry: "further" above)

**Reply by hishamco** [reply]
I didn't expect that both versions will be unpublish!!

**Reply by ns8482e** [reply]
`context.Cancel` should not change previously published content

**Comment by intellimedhu** [comment]
Deriving from a fatal exception solves the problem, but you may have a better solution.

**Comment by MikeAlhayek** [comment, marked as answer]
I think you're mixing between validation "which should be done in the driver" and handler. As suggested you could use `Context.Cancel` but that does not provide the user with errors or warnings so it could be misleading. Instead, you could use drivers to validate and raise errors based on your needs. Then in the handler so other work. If in the handler you used `Context.Cancel` and added error to the `ModelState`, the request will not show any error as far as I know. `Context.Cancel` will only prevent the data from being saved to the database but it will allow the request to proceed with no errors

**Reply by hishamco** [reply]
> Context.Cancel will only prevent the data from being saved to the database but it will allow the request to proceed with no errors

As he mentioned earlier that `Context.Cancel` is unpublish the current and published version which is weird

**Reply by MikeAlhayek** [reply]
@hishamco  The [Context.Cancel](https://github.com/OrchardCMS/OrchardCore/blob/aa2650522e9e57ede4256e3f270450d64414e7a2/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L320) just bails out and does not save the content to the session.. unless you are manually calling `_session.Save()` it would not publish or unpublish.

**Comment by intellimedhu** [comment]
Ok, this is a validation process must not be placed into a handler. I'll restructure my code. Thank you!
